tt_eager/tensor/tensor.cpp

```diff
@@ -186,11 +185,12 @@ void Tensor::deallocate(bool force) {
                                 [force, attr = this->tensor_attributes](Device* worker) mutable {
                                     ZoneScopedN("ShardDeallocate");
                                     auto& s = std::get<MultiDeviceStorage>(attr->storage);
-                                    if (s.buffers.find(worker->id()) != s.buffers.end()) {
-                                        if ((force or s.buffers.at(worker->id()).use_count() == 1)) {
-                                            DeallocateBuffer(*(s.buffers.at(worker->id())));
+                                    if (s.has_buffer_for_device(worker)) {
+                                        auto& device_buffer = s.get_buffer_for_device(worker);
```

Comment:
Another thread could remove this worker's buffer between these calls.

Reply:
Hmm, I'm not sure it can, unless I'm missing something. We only remove/reset buffers from storage objects inside the deallocate shard lambda function, which should only be called by the worker using this buffer.

Reply:
I see. My comment is probably not relevant for the PR but I worry that the complexity of which thread is allowed to modify which shared buffer when, without locking, is going to make code changes very difficult. I want to keep this in mind for the refactor.

Reply:
The scheme we've followed for all of multi-device async has been that the worker for device _i_ is the only one that can modify the buffer on device _i_ for a multi-device tensor. Its a 1:1 mapping.